### DISCUSSION TITLE
Getting the width and height of a page?

### DISCUSSION THREAD

**Discussion by PoshoDev** [zauberzeug/nicegui#2405] [Q&A] [answered]
### Question

Hello there, 🙂

I am trying to get the width and height of the whole page for something I'm working on. This seemed like a fairly simple thing to do, but I'm ashamed to say it has been quite a hard ordeal for me.

I've already tried the following:
- Researched the code for `ui.header` and `ui.left_drawer` (since it automatically hides when the page size becomes portrait, but it seems like the Quasar's `"show-if-above"` prop handles that logic).
- Looked into the `ui.aggrid`'s `ag-grid-community.min.js` code (since it also handles resizing, but that's too much for me to make any sense of).
- Investigating if such variables are hidden somewhere in the `Context`, `Page`, `Core`, `App` classes.
- Playing around with Tailwind and CSS (with no avail, due to inexperience).

The closest I've come is by rewriting @falkoschindler's solution from #1731 like this:
```python
with ui.page_sticky():#.props("hidden"):
    btn = ui.button().props("size=1px; dense").on("resize", lambda e: print(e.args))
    ui.add_head_html(f'''
    <script>
    function emitResizeEvent() {{
        content = document.querySelector('.nicegui-content');
        const dimensions = {{
            w: content.offsetParent.offsetWidth,
            h: content.offsetParent.offsetHeight
        }};
        getElement({btn.id}).$emit('resize', dimensions);
    }}
    window.onload = emitResizeEvent;
    window.onresize = emitResizeEvent;
    </script>
    ''')
```

In this code, I:
- Create an `ui.page_sticky` with an `ui.button` inside of it and put it in the bottom-right corner of the page.
- Create a modified version of the original `<script>` where I define a custom `.on("resize")` event and send the `content.offsetParent`'s `offsetWidth` and `offsetHeight` through a `dict` when the event happens.

This _kind_ of works, but has caveats:
- The user _needs_ to resize the page in order to retrieve the values. Ideally, I would need to know the values when the page loads.
- The values are inaccurate since the `ui.button` has some leftover size.

At this point I don't know what else to do but to ask for help. I'm going to cry when someone inevitably points out there was some sort of `page.get_width()` function all along and I've been dabbling with this for hours. 😭

**Comment by falkoschindler** [comment, marked as answer]
Hi @PoshoDev,

In version 1.4.7 we introduced a simpler way to emit global events:

```py
ui.add_head_html('''
    <script>
    function emitSize() {
        emitEvent('resize', {
            width: document.body.offsetWidth,
            height: document.body.offsetHeight,
        });
    }
    window.onload = emitSize;
    window.onresize = emitSize;
    </script>
''')

ui.on('resize', lambda e: print(f'resize: {e.args}'))
```

In my tests this emits the size when the page loads as well as when it is resized. Does this fulfil your needs?

**Reply by PoshoDev** [reply]
Thank you so very much! :smiley: 

This is exactly what I needed and is a billion times cleaner. You are a life saver!

**Reply by brunnenhof** [reply]
The above works perfectly, but I am trying to access the e.args (not just print them) to use the size programmatically further on in my python code. How do I get the values into python variables? 
Apologies for the dumb question ... Coming from Streamlit and Anvil.works to NiceGUI with next to zero JS knowledge.

**Reply by falkoschindler** [reply]
> How do I get the values into python variables?

This question is tricky to answer. You already _have_ the values. They are stored in `e.args`. What do you want to do? Usually you would call a function (like `print`) and pass `e.args` (or parts of it) as an argument.

Maybe this explanation from GPT-5 can help to clarify the difference between NiceGUI and Streamlit:

---

You’re running into a mental model mismatch with Streamlit.

**NiceGUI is not Streamlit.**

* Streamlit re-executes your script top-to-bottom on each interaction to rebuild the UI.
* NiceGUI starts a web server, builds the UI **once per client connection**, and then updates the existing DOM **only when events occur** (button clicks, timers, websockets messages, etc.). The whole script is **not** continuously re-evaluated.

Think of NiceGUI as **event-driven**:

* Top-level code sets up routes/pages and starts the server.
* When a user opens a page, the page function runs to create that user’s view.
* UI changes happen inside callbacks (on_click, on_change, timers, etc.) where you mutate widget properties or data. No global reruns.

Minimal example to show the difference:

```python
from nicegui import ui

state = {'count': 0}

with ui.row():
    label = ui.label(f'Count: {state["count"]}')
    def inc():
        state['count'] += 1
        label.text = f'Count: {state["count"]}'  # update just this widget
    ui.button('Increment', on_click=inc)

ui.run()
```

Notes:

* The file is executed once to start the app; it does **not** loop.
* Each client gets a live UI backed by callbacks.
* To update the UI, either mutate widget properties (like `label.text = ...`) in event handlers or use bindings/watchers if you prefer reactive patterns.

If you’re coming from Streamlit, a good way to translate your code is:

1. Put any “display” code inside a page function (or top-level if it’s a single page).
2. Move any code that “reacts” to user input into **callbacks**.
3. Update widget properties or your own state objects inside those callbacks (no need to re-run the entire script).

If you share a snippet of what you were trying in Streamlit terms, we can show the equivalent event-driven pattern in NiceGUI.

**Reply by brunnenhof** [reply]
> What do you want to do?

Thanks for the really fast reply - much appreciated! I want to use the size parameters to lay out input knobs (that control how much of a policy users want to implement), headers and matplotlib (or plotly, still pondering) output graphs differently on a mobile, tablet and desktop screen (and when users themselves resize the browser window).

Snippet in Streamlit: Don't have one, at the time I decamped to anvil.works due to their (lack of) state management in Streamlit. In Anvil.Works, the adjustment to screens and manual resizing seemed to happen 'magically' by itself.

I will sleep on the mindset advice and try again with a fresh outlook tomorrow morning.
